design/one-pager-change-logs.md

```diff
@@ -0,0 +1,410 @@
+  details through a
+  [`DeploymentRuntimeConfig`](https://docs.crossplane.io/latest/concepts/providers/#runtime-configuration)
+  resource. The specific details provided by the user are merged in with the
+  defaults from the package manager for a complete runtime specification.
+
+The `DeploymentRuntimeConfig` option is appealing, because it requires no code
+changes to core Crossplane and its package manager. At least while this feature
+is in Alpha, we will require users that want to enable change logs to do so via
+a `DeploymentRuntimeConfig` resource. A single `DeploymentRuntimeConfig` can be
+reused to configure multiple providers.
+
+Below is an example, informed by recent prototyping efforts, of what a
+`DeploymentRuntimeConfig` looks like that enables change logs for a provider.
+Note that it configures the sidecar container, mounts an `emptyDir` shared
+volume between the main provider and sidecar containers, and enables the alpha
+feature with the `--enable-change-logs` flag.
+
+```yaml
+apiVersion: pkg.crossplane.io/v1beta1
+kind: DeploymentRuntimeConfig
+metadata:
+  name: enable-change-logs
+spec:
+  deploymentTemplate:
+    spec:
+      selector: {}
+      template:
+        spec:
+          containers:
+          - name: package-runtime
+            args:
+            - --enable-change-logs
+            volumeMounts:
+            - name: change-log-vol
+              mountPath: /var/run/change-logs
+          - name: change-log-sidecar
+            image: crossplane/change-log-sidecar:latest
+            volumeMounts:
+            - name: change-log-vol
+              mountPath: /var/run/change-logs
+          volumes:
+          - name: change-log-vol
+            emptyDir: {}
+```
+
+### Implementation Rollout
+
+The general sequence of implementation work to roll this feature out is
+described below.  As previously mentioned, the plan does not include any updates
+to core Crossplane to support this feature, as we will rely on a
+`DeploymentRuntimeConfig` to perform the injection of the sidecar container and
+the setup of a shared volume between the main provider and sidecar containers.
+The key code changes we will make are in `crossplane-runtime`, a new
+`change-logs-sidecar` repo, the Upjet framework and template, and the ecosystem
+of providers.
+
+* `crossplane-runtime`:
+  * Protobuf types are defined for change log entries and a gRPC client/server
+    to send these entries from client to server
+  * Managed reconciler is updated to generate change log entries, send them to
+  the gRPC server (i.e. sidecar container), and to include additional calls to
+  `Observe()` after each CUD operation to capture the "after" state of the entry
+* `change-logs-sidecar`
+  * A new repository is created to build and publish the image used in the sidecar container
+  * Consumes the protobuf types from `crossplane-runtime`
+  * Starts the gRPC server and listens for change log entries
+  * Writes all entries to `stdout` (i.e. pod logs)
```

Comment:
Interesting idea @turkenh - in general I like the idea of using standard tools to not have to build too much ourselves and to integrate with other systems more easily.

Do you know of any examples that show how to actually use the logs SDK? Just reading through the [spec](https://opentelemetry.io/docs/specs/otel/logs/sdk/) you linked, I'm already confused about how to actually make it work 😂. I'm not quickly finding practical examples of how to set up and configure these components, perhaps you already know where to find those? 

I've basically built the entire scenario described in this design document end to end already, see:
* https://github.com/crossplane/crossplane-runtime/compare/master...jbw976:crossplane-runtime:change-logs
* https://github.com/crossplane-contrib/provider-kubernetes/compare/main...jbw976:provider-kubernetes:change-logs
* https://github.com/jbw976/change-log-sidecar
  * a lot of these files are just for my dev/testing iterations 😇 

The code isn't too complex (but certainly needs polishing too), so the maintenance overhead over time isn't too daunting. I also like how the experience is very specific and scoped to the information we're trying to collect from crossplane-runtime and the providers, as opposed to getting something more heavy weight involved.

Your last 2 points are probably the most interesting (about being able to easily send to a central instance, and to integrate with processors easily). I had assumed that if someone chooses to set up the otel collector in their control plane, and have it easily gather the change log sidecar logs which are nicely isolated for them, they could use processors too. Do you think that isn't possible unless the log entries were originally emitted with the otel logs SDK?

How strongly do you feel about this idea? 🤔

Reply:
I looked into the Otel logging approach before I suggested the design that @jbw976 has been running with. While I'd typically agree it's best not to reinvent the wheel,  I wasn't convinced we should use Otel logs (or really events) for this.

Quoting myself from Upbound's internal Slack a few weeks ago:

> I looked into otel (log) events a bit more this afternoon, as a possible way to emit change logs:
> 
> * https://opentelemetry.io/docs/specs/otel/logs/event-api/#emit-event
> * https://github.com/open-telemetry/opentelemetry-specification/issues/4045
> 
> They’re a bit oddball. At the RPC layer they’re just a log with no severity, an event.name attribute, and (by  convention) a structured body , with the same schema for all events with the same name. The Go SDK only has beta support for logging and has no explicit support for events (SDKs are supposed to have specific abstractions for events even though they’re just specially formatted logs). You could just manually emit logs that are by convention an event, but at the end of the day the SDKs just wrap some underlying logger (e.g. zap).

AFAICT:

* The Go logs SDK seems fairly immature. It's beta, and subject to breaking changes.
* We really want events (a special type of log), and there is no Go events SDK yet.
* The "structured" body of an Otel log isn't really structured - it's just an `Any` - https://github.com/open-telemetry/opentelemetry-proto/blob/v1.3.2/opentelemetry/proto/logs/v1/logs.proto#L170

I don't feel very strongly, but my intuition is that there's limited benefit to using Otel for this and that trying to fit into their model might make like more difficult than it needs to be.

Reply:
> One can easily configure pushing logs to a central/shared otel collector instead, avoiding 1 sidecar per provider.

Whether we use Otel or not, I'd prefer to avoid _pushing_ logs to a centralized collector.

Pushing to a sidecar container was an intentional design choice. It gives us a local(ish) cache of change logs that a centralized collector can scrape. If we were to push logs to a centralized collector we'd risk losing change logs if a provider was partitioned from the collector. @jbw976 touches on this under the "Storage and Durability" section of the design.